File: market-basket-analysis/src/main/scala/com/my/mba/spark/AssociationRuleApp.scala
package com.my.mba.spark

import org.apache.hadoop.fs.{FileSystem, Path}
import org.apache.spark.{SparkConf, SparkContext}

import scala.collection.mutable.ListBuffer

/**
  * scala版计算关联规则
  */
object AssociationRuleApp {
  val basePath = "data"
  val support = 0.00
  val confidence = 0.00
  val delimiter = ","

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("AssociationRuleApp")
    conf.setMaster("local[*]")
    val sc = new SparkContext(conf)
    val fileRDD = sc.textFile(s"${basePath}/test/test.csv", 1)
    val txnCount = fileRDD.count()

    // 获得所有候选项集
    val itemsCountRDD = fileRDD.flatMap(line => {
      val items = line.split(delimiter)
      (0 to items.size) flatMap items.combinations filter (xs => !xs.isEmpty)
    }).map((_, 1))
    val minSupportCount = support * txnCount
    //获得所有频繁项集
    val frequentRDD = itemsCountRDD.reduceByKey(_ + _).filter(_._2 >= minSupportCount)

    val subItemSetRDD = frequentRDD.flatMap {
      case (frequentItemSet, supportCount) =>
        val result = ListBuffer.empty[(List[String], (List[String], Int))]
        result += ((frequentItemSet, (Nil, supportCount)))
        //即把K作为K2，Tuple(null,V))作为V2
        val sublist =
          (0 to frequentItemSet.size - 1) flatMap frequentItemSet.combinations filter (xs => !xs.isEmpty) map((_, (frequentItemSet, supportCount)))
        result ++= sublist
        result.toList
    }
    val rules = subItemSetRDD.groupByKey()
    val associationRulesRDD = rules.flatMap { case (antecedent, allSuperItemSets) =>
      val antecedentSupportCount = allSuperItemSets.find(p => p._1 == Nil).get._2
      val superItemSets = allSuperItemSets.filter(p => p._1 != Nil).toList //将规则集合去掉空的
      if (superItemSets.isEmpty) Nil
      else {
        val result =
          for {
            (superItemSet, superItemSetSupportCount) <- superItemSets
            consequent = superItemSet diff antecedent
            ruleSupport = superItemSetSupportCount.toDouble / txnCount.toDouble
            confidence = superItemSetSupportCount.toDouble / antecedentSupportCount.toDouble if confidence >= minConfidence
          } yield (((antecedent, consequent, ruleSupport, confidence)))
        result
      }
    }
    val formatResult = associationRulesRDD.map(s => {
      (s._1.mkString("[", ",", "]"), s._2.mkString("[", ",", "]"), s._3, s._4)
    })

    FileSystem.get(sc.hadoopConfiguration).delete(new Path(output), true)
    formatResult.saveAsTextFile(output)
    sc.stop()
  }

}

File: market-basket-analysis/src/main/java/com/my/mba/spark/AssociationRuleApp.java
package com.my.mba.spark;

import com.my.mba.util.Combination;
import com.my.mba.util.Utils;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFlatMapFunction;
import scala.Tuple2;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * spark计算关联规则
 */
public class AssociationRuleApp {
    public static void main(String[] args) throws IOException {
        final double minSupport = 0.40;
        final double minConfidence = 0.60;
        String delimiter = ",";
        SparkConf conf = new SparkConf();
        conf.setAppName("AssociationRuleApp");
        conf.setMaster("local[*]");
        JavaSparkContext context = new JavaSparkContext(conf);
        JavaRDD<String> fileRDD = context.textFile("F:\\BigData\\workspace\\spark-demo\\market-basket-analysis\\src\\main\\resources\\test\\test.csv", 1);
      //1. 求出原始数据的总大小
        long  txnCount = fileRDD.count();
        System.out.println("txnCount=" + txnCount);
        double minSupportCount = txnCount * minSupport;
        //2. 计算频繁项集
        JavaPairRDD<List<String>, Integer> itemSetCountRDD = fileRDD.flatMapToPair(new PairFlatMapFunction<String, List<String>, Integer>() {
            @Override
            public Iterator<Tuple2<List<String>, Integer>> call(String s) throws Exception {
                List<String> items = Utils.toList(s, delimiter);
                List<List<String>> subItemsSets = Combination.findSortedCombinations(items);
                List<Tuple2<List<String>, Integer>> result = new ArrayList<Tuple2<List<String>, Integer>>();
                for (List<String> subItemsSet : subItemsSets) {
                    if(subItemsSet.size() > 0)
                    result.add(new Tuple2<List<String>, Integer>(subItemsSet, 1));
                }

                return result.iterator();
            }
        });

        JavaPairRDD<List<String>, Integer> frequentItemSetRDD = itemSetCountRDD.reduceByKey(new Function2<Integer, Integer, Integer>() {
            @Override
            public Integer call(Integer v1, Integer v2) throws Exception {
                return v1 + v2;
            }
        }).filter(new Function<Tuple2<List<String>, Integer>, Boolean>() {
            @Override
            public Boolean call(Tuple2<List<String>, Integer> v1) throws Exception {
                return v1._2() >= minSupportCount;
            }
        });

        //3. 根据频繁项集计算出关联规则
        JavaPairRDD<List<String>, Tuple2<List<String>, Integer>> subItemSetRDD = frequentItemSetRDD.flatMapToPair(
                new PairFlatMapFunction<Tuple2<List<String>, Integer>, List<String>, Tuple2<List<String>, Integer>>() {
                    @Override
                    public Iterator<Tuple2<List<String>, Tuple2<List<String>, Integer>>> call(
                            Tuple2<List<String>, Integer> frequentItemSet) {
                        List<Tuple2<List<String>, Tuple2<List<String>, Integer>>> result = new ArrayList<Tuple2<List<String>, Tuple2<List<String>, Integer>>>();
                        List<String> itemSet = frequentItemSet._1;
                        Integer itemSetSupportCount = frequentItemSet._2;
                        result.add(new Tuple2(itemSet, new Tuple2(null, itemSetSupportCount))); // 主要是为了后面区分开，好拿到父集出现的次数
                        if (itemSet.size() == 1) {
                            return result.iterator();
                        }
                        List<List<String>> allSubItemSets = Combination.findSortedCombinations(itemSet);
                        allSubItemSets.remove(0);
                        allSubItemSets.remove(allSubItemSets.size() - 1);
                        for(List<String> subItemSet : allSubItemSets) {
                            result.add(new Tuple2<List<String>, Tuple2<List<String>, Integer>>(subItemSet,
                                    new Tuple2(itemSet, itemSetSupportCount)));
                        }
                        return result.iterator();
                    }
                });

        JavaPairRDD<List<String>, Iterable<Tuple2<List<String>, Integer>>> rules = subItemSetRDD.groupByKey();

        JavaRDD<String> associationRulesRDD =
                rules.flatMap(new FlatMapFunction<Tuple2<List<String>, Iterable<Tuple2<List<String>, Integer>>>, String>() {
                    @Override
                    public Iterator<String> call(Tuple2<List<String>, Iterable<Tuple2<List<String>, Integer>>> rule) throws Exception {
                        List<String> result = new ArrayList<String>();
                        List<String> antecedent = rule._1;
                        Iterable<Tuple2<List<String>, Integer>> inSuperItemSets = rule._2;
                        List<Tuple2<List<String>, Integer>> superItemSets = new ArrayList<Tuple2<List<String>, Integer>>();
                        Integer antecedentSupportCount = 0;
                        for (Tuple2<List<String>, Integer> t2 : inSuperItemSets) {
                            // find the "count" object
                            if (t2._1 == null) {
                                antecedentSupportCount = t2._2;
                            } else {
                                superItemSets.add(t2);
                            }
                        }
                        if (superItemSets.isEmpty()) {
                            return result.iterator();
                        }
                        for (Tuple2<List<String>, Integer> t2 : superItemSets) {
                            // 计算 后件 = superItemSet - 前件
                            List<String> consequent = new ArrayList<String>(t2._1);
                            consequent.removeAll(antecedent);

                            double ruleSupport = (double) t2._2 / (double)txnCount;
                            double confidence =  (double) t2._2 / (double)antecedentSupportCount;

                            if (confidence >= minConfidence && !antecedent.isEmpty() && !consequent.isEmpty()) {
                                result.add("(" + antecedent + "=>" + consequent + "," + ruleSupport + "," + confidence + ")");
                            }
                        }
                        return result.iterator();
                    }
                });

        String outputPath  = "F:\\BigData\\workspace\\spark-demo\\data\\spark";
        FileSystem.get(context.hadoopConfiguration()).delete(new Path(outputPath), true);
        associationRulesRDD.saveAsTextFile(outputPath);
        context.stop();
    }
}
